tensorflow/compiler/xla/service/hlo_instructions.h

```diff
@@ -1217,8 +1217,14 @@ class HloCustomCallInstruction : public HloInstruction {
   void set_batch_group_count(int64 batch_group_count) {
     batch_group_count_ = batch_group_count;
   }
+  // Sets whether this custom call has a side-effect - by default a custom call
+  // has no side-effects.
+  void set_has_side_effect(bool has_side_effect) {
+    has_side_effect_ = has_side_effect;
+  }
   int64 feature_group_count() const { return feature_group_count_; }
   int64 batch_group_count() const { return batch_group_count_; }
+  bool has_side_effect() const { return has_side_effect_; }
```

Comment:
Since we already have `HloInstruction::HasSideEffect`, I'd prefer not to have this function at all.

I guess you need it because you want `HloInstruction::HasSideEffect` to *call* this.  But maybe we just make HloInstruction a friend of HloCustomCallInstruction and have it read the bit directly?  That's kind of ugly (really should have that virtual function :-p) but I think it's better than having both `has_side_effect()` and `HasSideEffect()`?

Reply:
I'm not sure what the rules are, but if I were making them then one would be 'never use friend unless it is for letting a unit test peer inside its testee'.

if HasSideEffect is the main query function, can we have HaSideEffect virtual, and have the HloCustomClassInstruction implementation of it call down to the HloInstruction version and then OR in it's own class local state?

Reply:
alternatively rename 'has_side_effect' to 'custom_call_has_side_effect' ?

Reply:
The original proposal was to make HloInstruction have a virtual HasSideEffect but nobody would actually override it, in the original patch.  I'd be OK making HasSideEffect virtual *and using it in HloCustomCallInstruction*.  There's still a speed hit to making HasSideEffect virtual for only one edge case that I'm not thrilled about, but at least we'd be using the extension point we're adding.

I'm also OK with `custom_call_has_side_effect`.

Using `friend` doesn't bother me so much because these classes are all a mess, stuck in the middle of an uncompleted refactoring.  :(